src/net.cpp

```diff
@@ -420,6 +420,26 @@ CNode* ConnectNode(CAddress addrConnect, const char *pszDest, bool fConnectToMas
             return NULL;
         }
 
+        if (pszDest && addrConnect.IsValid()) {
+            // It is possible that we already have a connection to the IP/port pszDest resolved to.
+            // In that case, drop the connection that was just created, and return the existing CNode instead.
+            // Also store the name we used to connect in that CNode, so that future FindNode() calls to that
+            // name catch this early.
+            CNode* pnode = FindNode((CService)addrConnect);
+            if (pnode)
+            {
+                pnode->AddRef();
```

Comment:
I'm thinking about backporting bitcoin/bitcoin#9626 that in both cases replaces
```
pnode->AddRef();
return pnode;
```
with
```
LogPrintf("Failed to open new connection, already connected\n");
return NULL;
```
So, looks to me that after this backport, upgrading node's status to masternode can be done with one line, like
```
pnode->fMasternode = pnode->fMasternode || fConnectToMasternode;
```
BTW, what was the reason to do `pnode->AddRef();` in this place only when upgrading node's status to masternode, not always (like in bitcoin).